src/v/kafka/server/handlers/metadata.cc

```diff
@@ -269,22 +269,113 @@ get_topic_metadata(request_context& ctx, metadata_request& request) {
+    // (https://github.com/vectorizedio/redpanda/issues/3588)
+    //
+    // Use a fallback matching to find the best peer address we can.
+    vlog(
+      klog.warn,
+      "Broker {} has no listener named '{}', falling "
+      "back to guessing peer listener",
+      broker->id(),
+      ctx.listener());
+
+    // Look up port for the listener in use for this request
+    const auto& my_listeners = config::node().advertised_kafka_api();
+    int16_t my_port = 0;
+    for (const auto& l : my_listeners) {
+        if (l.name == ctx.listener()) {
+            my_port = l.address.port();
+
+            // Looking up the address for myself?  Take the whole
+            // listener.  This is the path where what's in node_config
+            // is not yet consistent with what's in members_table,
+            // because a node configuration update didn't propagate
+            // via raft0 yet
+            if (broker->id() == config::node().node_id()) {
+                return &l;
+            }
+        }
+    }
+
+    if (my_port == 0) {
+        // Should never happen: if we're listening with a given
+        // name, that name must have been in config.
+        vlog(
+          klog.error,
+          "Request on listener '{}' but not found in node_config",
+          ctx.listener());
+        return nullptr;
+    }
+
+    // Fallback 1: Try to match by port
+    for (const auto& listener : broker->kafka_advertised_listeners()) {
+        // filter broker listeners by active connection
+        if (listener.address.port() == my_port) {
+            return &listener;
+        }
+    }
+
+    // Fallback 2: no name or port match, return first listener from
+    // peer.
+    if (!broker->kafka_advertised_listeners().empty()) {
+        return &broker->kafka_advertised_listeners()[0];
+    } else {
+        // A broker with no kafka listeners, there is no way to
+        // include it in our response
+        return nullptr;
+    }
+}
+
 template<>
 ss::future<response_ptr> metadata_handler::handle(
   request_context ctx, [[maybe_unused]] ss::smp_service_group g) {
     metadata_response reply;
     auto alive_brokers = co_await ctx.metadata_cache().all_alive_brokers();
     for (const auto& broker : alive_brokers) {
+        const model::broker_endpoint* peer_listener = nullptr;
         for (const auto& listener : broker->kafka_advertised_listeners()) {
             // filter broker listeners by active connection
             if (listener.name == ctx.listener()) {
-                reply.data.brokers.push_back(metadata_response::broker{
-                  .node_id = broker->id(),
-                  .host = listener.address.host(),
-                  .port = listener.address.port(),
-                  .rack = broker->rack()});
+                peer_listener = &listener;
+                break;
             }
         }
+
+        if (peer_listener == nullptr) {
+            peer_listener = guess_peer_listener(ctx, broker);
+        }
```

Comment:
I really believe that I reviewed this same code but in a different PR earlier today. Am I losing it?

Reply:
Hah, no, you reviewed the admin API redirect code, which also has a "guess" function for inferring a peer's admin address.

Reply:
Ah ha!